syft/sbom/format.go

```diff
@@ -21,88 +13,19 @@ func (f FormatID) String() string {
 
 const AnyVersion = ""
 
-type Format interface {
+type FormatEncoder interface {
 	ID() FormatID
-	IDs() []FormatID
+	Aliases() []string
 	Version() string
 	Encode(io.Writer, SBOM) error
-	Decode(io.Reader) (*SBOM, error)
-	Validate(io.Reader) error
-	fmt.Stringer
-}
-
-type format struct {
-	ids       []FormatID
-	version   string
-	encoder   Encoder
-	decoder   Decoder
-	validator Validator
-}
-
-func (f format) IDs() []FormatID {
-	return f.ids
-}
-
-func (f format) Version() string {
-	return f.version
-}
-
-func (f format) String() string {
-	if f.version == AnyVersion {
-		return f.ID().String()
-	}
-	return fmt.Sprintf("%s@%s", f.ID(), f.version)
-}
-
-// Decoder is a function that can convert an SBOM document of a specific format from a reader into Syft native objects.
-type Decoder func(reader io.Reader) (*SBOM, error)
-
-// Encoder is a function that can transform Syft native objects into an SBOM document of a specific format written to the given writer.
-type Encoder func(io.Writer, SBOM) error
-
-// Validator reads the SBOM from the given reader and assesses whether the document conforms to the specific SBOM format.
-// The validator should positively confirm if the SBOM is not only the format but also has the minimal set of values
-// that the format requires. For example, all syftjson formatted documents have a schema section which should have
-// "anchore/syft" within the version --if this isn't found then the validator should raise an error. These active
-// assertions protect against "simple" format decoding validations that may lead to false positives (e.g. I decoded
-// json successfully therefore this must be the target format, however, all values are their default zero-value and
-// really represent a different format that also uses json)
-type Validator func(reader io.Reader) error
-
-func NewFormat(version string, encoder Encoder, decoder Decoder, validator Validator, ids ...FormatID) Format {
-	return format{
-		ids:       ids,
-		version:   version,
-		encoder:   encoder,
-		decoder:   decoder,
-		validator: validator,
-	}
 }
 
-func (f format) ID() FormatID {
-	return f.ids[0]
-}
+type FormatDecoder interface {
+	// Decode will return an SBOM from the given set of bytes. If the bytes are not a valid SBOM for the given format
+	// then an error will be returned.
+	Decode(by []byte) (*SBOM, FormatID, string, error)
 
-func (f format) Encode(output io.Writer, s SBOM) error {
-	if f.encoder == nil {
-		return ErrEncodingNotSupported
-	}
-	return f.encoder(output, s)
+	// Identify will return the format ID and version for the given set of bytes. Note: this does not validate the
+	// full SBOM, only pulls the minimal information necessary to identify the format.
+	Identify(by []byte) (FormatID, string)
```

Comment:
I can take a second pass at this -- I tried preserving this behavior but realized that most implementations needed the entire doc to function anyway. Necessarily it would need to be a ReadSeeker since multiple passes are done.

Reply:
luckily we're already smuggling ReadSeeker behavior in the underlying objects via stereoscope, so moving to ReadSeekers should be viable 👍  -- I'll head that direction. I'll be adding a lot of seek-0-just-in-case calls to ensure we don't get subtle bugs when using decoders within a collection.